nixos/gdm: Fix fingerprint auth rules

jtojnar opened the pull request:
## Description of changes

We introduced the gdm-fingerprint.pam in 9d41fe6fcc4df838a56b1cfb2512b65e5e655958.

We used the [upstream Arch config] as a template, which contains an extended control field that jumps over **one** immediately-following `auth` rule unless `pam_gdm.so` succeeds.

But we decided to not include `pam_gnome_keyring.so` so there was no rule to skip over, resulting in a broken control flow and the PAM module failing with “PAM bad jump in stack”, breaking the fingerprint authentication in GDM.

Let’s actually add `pam_gnome_keyring.so`, like the Arch config does. Because we are creating the PAM file using the `text` option, `security.pam.services.gdm-fingerprint.enableGnomeKeyring` does not do anything so we need to do it manually.

For the case where gnome-keyring is not enabled, we could add a no-op rule like `optional pam_permit.so` after `pam_gdm.so` so that the branching always has something to jump over but it will be simpler to just make the both conditional. There are no further `auth` rules that could benefit from `pam_gdm.so` doing something so it should be fine.

Unlike in Arch, we are not going to invoke `pam_gnome_keyring.so` in a `session` rule since that is already done by the included `login` module.

Let’s also add a comment about disabling `fprintAuth`, remove no-op `optional pam_permit.so` rule and revert parts of https://github.com/NixOS/nixpkgs/pull/315874 that do not do anything.

[upstream Arch config]: https://gitlab.gnome.org/GNOME/gdm/-/blob/81ee658c11381912131dd4a29e84190f7f9cd039/data/pam-arch/gdm-fingerprint.pam


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jtojnar committed:
nixos/gdm: Fix fingerprint auth rules

We introduced the gdm-fingerprint.pam in 9d41fe6fcc4df838a56b1cfb2512b65e5e655958.

We used the [upstream Arch config] as a template, which contains an extended control field that jumps over **one** immediately-following `auth` rule unless `pam_gdm.so` succeeds.

But we decided to not include `pam_gnome_keyring.so` so there was no rule to skip over, resulting in a broken control flow and the PAM module failing with “PAM bad jump in stack”, breaking the fingerprint authentication in GDM.

Let’s actually add `pam_gnome_keyring.so`, like the Arch config does. Because we are creating the PAM file using the `text` option, `security.pam.services.gdm-fingerprint.enableGnomeKeyring` does not do anything so we need to do it manually.

For the case where gnome-keyring is not enabled, we could add a no-op rule like `optional pam_permit.so` after `pam_gdm.so` so that the branching always has something to jump over but it will be simpler to just make the both conditional. There are no further `auth` rules that could benefit from `pam_gdm.so` doing something so it should be fine.

Unlike in Arch, we are not going to invoke `pam_gnome_keyring.so` in a `session` rule since that is already done by the included `login` module.

[upstream Arch config]: https://gitlab.gnome.org/GNOME/gdm/-/blob/81ee658c11381912131dd4a29e84190f7f9cd039/data/pam-arch/gdm-fingerprint.pam

Pandapip1 commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -321,21 +322,31 @@ in
-        auth       optional                    pam_permit.so
         auth       required                    pam_env.so
         auth       [success=ok default=1]      ${pkgs.gnome.gdm}/lib/security/pam_gdm.so
+        ${lib.optionalString
+          pamCfg.gdm-fingerprint.enableGnomeKeyring
+          "auth       optional                    ${pkgs.gnome-keyring}/lib/security/pam_gnome_keyring.so"}
+        # So that pam_gdm has something to skip over in the default branch when gnome-keyring is disabled.
+        auth       optional                    pam_permit.so
```
```suggestion
        ${if pamCfg.gdm-fingerprint.enableGnomeKeyring then
          "auth       optional                    ${pkgs.gnome-keyring}/lib/security/pam_gnome_keyring.so" else
          # So that pam_gdm has something to skip over in the default branch when gnome-keyring is disabled.
          "auth       optional                    pam_permit.so"}
```
So that the `pam_permit.so` line is only included when it needs to be.

jtojnar replied:
Hmm, I am split on whether that is easier to read. But it gave me an idea – I wonder if it would make sense to remove the `pam_permit.so` altogether and use something like the following control for `pam_gdm.so`:

```nix
${if pamCfg.gdm-fingerprint.enableGnomeKeyring then "[success=ok default=1]" else "optional"}
```

Pandapip1 replied:
I agree, that's preferable.

jtojnar replied:
Decided to make both lines conditional instead, since that will be even cleaner. There are no further `auth` rules that could benefit from `pam_gdm.so` doing something.

Pandapip1 reviewed (approved):
LGTM; one nitpick.

JohnRTitor commented:
We introduced `gdm-password.enableGnomeKeyring` in #315874 as well, can we make its PAM config more in line with [upstream](https://gitlab.gnome.org/GNOME/gdm/-/blob/main/data/pam-arch/gdm-password.pam)? 
https://github.com/NixOS/nixpkgs/blob/6cb20d80466067bc39c00a02d28f866f0db05965/nixos/modules/services/x11/display-managers/gdm.nix#L304-L309

jtojnar commented:
Good point, `gdm-password.enableGnomeKeyring = true;` and `gdm-autologin.enableGnomeKeyring = true;` won’t do anything either.

Though, I do not think it makes sense to copy the Arch PAM config as-is everywhere – we already have `pam_gnome_keyring.so` in `login` PAM module.

Actually, maybe, thanks to that, we might not need `pam_gnome_keyring.so` here at all.

I believe Arch skips the `pam_gnome_keyring.so` rule when `pam_gdm.so` fails. But since we were already calling `pam_gnome_keyring.so` in `pam.d/login` without issue even without running `pam_gdm.so`, it is probably fine to get rid of the conditional. After all, the only thing `pam_gdm.so` does is make sure passwords from early boot are available to later rules (same as `pam_systemd_loadkeys.so`, which should replace it). And I would expect the `optional pam_gnome_keyring.so` rule to just fail when a password is not available.

jtojnar committed:
Revert "nixos/gnome-keyring: enable gnome-keyring for fingerprint authentication'"

`gdm-fingerprint` PAM module is defined using the `text` option, so the option here is a no-op.

This reverts commit 6bb516d45f2cbb56a817adf4c7f0ee680e3cf9e9.

jtojnar committed:
Revert "nixos/gnome-keyring: unlock keyring with gdm-password, gdm-autologin"

`gdm-autologin` and `gdm-password` PAM modules are defined using the `text` option, so the option here is a no-op.

Furthermore, `gdm-password` already includes `login` for all module types,
and that invokes `pam_gnome_keyring.so` in the same way Arch’s `gdm-password` module would:
https://gitlab.gnome.org/GNOME/gdm/-/blob/81ee658c11381912131dd4a29e84190f7f9cd039/data/pam-arch/gdm-password.pam

This reverts commit c24c7933ba2e5266d91978f1eaefdd81c760af67.

jtojnar commented:
Actually, never mind – the `gdm-fingerprint.pam` does not contain `auth include login` so we still need to add `pam_gnome_keyring.so` ourselves.

I have mostly reverted https://github.com/NixOS/nixpkgs/pull/315874 since it does not do anything.

We should probably replace `gdm-autologin` with https://gitlab.gnome.org/GNOME/gdm/-/blob/81ee658c11381912131dd4a29e84190f7f9cd039/data/pam-arch/gdm-autologin.pam but I won’t do it here so that we do not block fingerprint support further.

jtojnar committed:
nixos/gdm: Clean up gdm-fingerprint pam module

The `optional pam_permit.so` comes from the [upstream Arch config] we used as a template in 9d41fe6fcc4df838a56b1cfb2512b65e5e655958. But I do not think it does anything in this position – see also the discussion at https://bbs.archlinux.org/viewtopic.php?id=245892 – so let’s just remove it.

Let’s also add a comment about disabling `fprintAuth` and a blank line for clarity.

[upstream Arch config]: https://gitlab.gnome.org/GNOME/gdm/-/blob/81ee658c11381912131dd4a29e84190f7f9cd039/data/pam-arch/gdm-fingerprint.pam

JohnRTitor reviewed (approved):
LGTM

schuelermine commented:
seems to work fine on my machine when cherrypicked onto nixos-unstable
